core-rust/state-manager/src/mempool/mempool_manager.rs

```diff
@@ -166,36 +163,28 @@ impl MempoolManager {
             .collect()
     }
 
-    /// Checks the committability of a random subset of transactions and removes the rejected ones
-    /// from the mempool.
+    /// Checks the committability of a subset of transactions executed against earliest state versions
+    /// and removes the newly rejected ones from the mempool.
     /// Obeys the given limit on the number of actually executed (i.e. not cached) transactions.
     pub fn reevaluate_transaction_committability(&self, max_reevaluated_count: u32) {
-        // TODO: better selection of transactions based on last time/state version against it was reevaluated.
-        const MIN_TRANSACTIONS_TO_CHECK_FOR_REEVALUATION: u32 = 100;
-        let candidate_transactions = self.mempool.read().get_k_random_transactions(max(
-            max_reevaluated_count,
-            MIN_TRANSACTIONS_TO_CHECK_FOR_REEVALUATION,
-        )
-            as usize);
+        let candidate_transactions: Vec<Arc<MempoolData>> = self
+            .mempool
+            .read()
+            .iter_by_state_version()
+            .take(max_reevaluated_count as usize)
+            .collect();
 
         let mut transactions_to_remove = Vec::new();
-        let mut reevaluated_count = 0;
         for candidate_transaction in candidate_transactions {
-            let (record, was_cached) = self
+            let (record, _was_cached) = self
                 .cached_committability_validator
                 .check_for_rejection_cached_prevalidated(
-                    &candidate_transaction.validated,
-                    ForceRecalculation::No,
+                    &candidate_transaction.transaction.validated,
+                    ForceRecalculation::Yes,
```

Comment:
I remember that 2 weeks ago I (thought I) knew why this change happened, but now I don't :D can you please confirm it is intentional?

Reply:
Yes, it is intentional. Before, the result cache didn't remove anything from here, hence why we preferred to only recalculate with the cache decided so. Now, before getting here, we remove from the mempool everything that is invalidated and everything left will be valid (for that state version) if we do not reexecute.